### PR TITLE
export opentelemetry spans with batch processor

### PR DESCRIPTION
**Pull request by theoclark**
We have observed slowdowns of up to 50% when running tritonserver with opentelemetry. We have traced this to the SimpleSpan Exporter and have observed these slowdowns go away when switching to the BatchSpanProcessor.

### CODE REVIEW

File: src/tracer.cc [hunk 1]
```diff
@@ -48,6 +48,9 @@ namespace otel_resource = opentelemetry::sdk::resource;
 
 namespace triton { namespace server {
 
+static const int EXPORT_QUEUE_SIZE = 5000;
```
Comment (oandreeva-nv) [inline src/tracer.cc:51]: Does it make sense to make these options configurable through cmdline args?

File: src/tracer.h [hunk 2]
```diff
@@ -40,6 +40,8 @@
 #include "opentelemetry/sdk/resource/resource.h"
 #include "opentelemetry/sdk/trace/processor.h"
 #include "opentelemetry/sdk/trace/simple_processor_factory.h"
```
Comment (oandreeva-nv) [inline src/tracer.h:42]: If we are removing `SimpleProcessFactory`, then it makes sense to remove this input

**Review by oandreeva-nv — COMMENTED** [review]
Hi @theoclark , thank you for this PR! Left some comments. May I also ask you to write some tests for your code? Open telemetry tests are located here: https://github.com/triton-inference-server/server/tree/main/qa/L0_trace.  Could you please also send us a signed [CLA](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla)?

**Comment by theoclark** [comment]
> Hi @theoclark , thank you for this PR! Left some comments. May I also ask you to write some tests for your code? Open telemetry tests are located here: https://github.com/triton-inference-server/server/tree/main/qa/L0_trace. Could you please also send us a signed [CLA](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla)?

Thanks for the review, will make some changes. I work for Speechmatics and I believe we've already signed the CLA? Any specific tests you had in mind?

**Comment by oandreeva-nv** [comment]
> Thanks for the review, will make some changes. I work for Speechmatics and I believe we've already signed the CLA? Any specific tests you had in mind?

Got it, thanks for letting me know about CLA. Regarding tests. Basically, it would be nice to add the following tests:
- Tests for setting new cmdline args. E.g. test for small queue size (3 or 4) and see if OTEL collector collects all 4 traces only after Triton receives 4 traces. Let me know if it makes sense. 
- Existing tests need adjustments. Since we are removing Simple Span Processor, current tests will fail, unless the default values for `BatchSpanProcessor` are set to imitate `SimpleSpanProcessor`.

**Comment by oandreeva-nv** [comment]
Hi @theoclark , I've pushed a PR, which has a major refactor for tests: https://github.com/triton-inference-server/server/pull/6785

If you are interested, I can pick up this PR as well and add tests myself.

**Comment by theoclark** [comment]
Hi @oandreeva-nv, thanks that would be great. Hadn't got around to writing the tests myself.

**Comment by theoclark** [comment]
@oandreeva-nv we've also noticed in other parts of our code that exporting traces across HTTP can have more of an impact on performance than gRPC. Is there a reason tritonserver uses HTTP over gRPC that I'm not aware of?

**Comment by oandreeva-nv** [comment]
@theoclark There was a limited scope of what we could implement initially. gRPC is on our roadmap for implementation. I'll try to make it happen in 24.02 release, since there is one more ask for gRPC exporter

**Comment by oandreeva-nv** [comment]
Hi @theoclark , I'll close this PR since https://github.com/triton-inference-server/server/pull/6842 was merged, and it includes your commits from this PR and mentions you as a co-author

**Comment by theoclark** [comment]
Thanks @oandreeva-nv for your work on this. Great that it's been merged.

**Pull request closed without merging**